File: presence/rust/engine/presence_cpp_ffi/src/rust_to_c.rs
// Construct data structures passed from Rust to C.
include!(concat!(env!("OUT_DIR"), "/rust_to_c_autogen.rs"));

use presence_core::ble_scan_provider::ScanRequest;
use presence_core::client_provider::{DiscoveryResult, PresenceMedium};

// Convert ScanRequest to *PresenceBleScanRequest, which can be parsed by C codes.
impl PresenceBleScanRequest {
    pub fn from_scan_request(scan_request: ScanRequest) -> *mut Self {
        unsafe {
            let ble_scan_request = presence_ble_scan_request_new(scan_request.priority);
            for action in scan_request.actions {
                presence_ble_scan_request_add_action(ble_scan_request, action);
            }
            ble_scan_request
        }
    }
}

// Convert DiscoveryResult to *PresenceDiscoveryResult, which can be parsed by C codes.
impl PresenceDiscoveryResult {
    pub fn from_discovery_result(result: DiscoveryResult) -> *mut Self {
        unsafe {
            let presence_result = presence_discovery_result_new(result.medium);
            for action in result.device.actions {
                presence_discovery_result_add_action(presence_result, action);
            }
            presence_result
        }
    }
}

File: presence/rust/engine/presence_cpp_ffi/cpp/rust_to_c.cpp
#include <stdio.h>
#include <stdlib.h>
#include "../presence.h"
#include "rust_to_c.hpp"

struct PresenceBleScanRequest* presence_ble_scan_request_new(int priority) {
  PresenceBleScanRequest* request = new PresenceBleScanRequest();
  request->priority = priority;
  return request;
}

void presence_ble_scan_request_add_action(PresenceBleScanRequest* request,
                                          int action) {
  request->actions.push_back(action);
}

void presence_ble_scan_request_free(struct PresenceBleScanRequest* request) {
  delete request;
}

struct PresenceDiscoveryResult* presence_discovery_result_new(enum PresenceMedium medium) {
  // struct PresenceDevice presence_device = { .actions = NULL, .actions_size = 20 };
  // PresenceDiscoveryResult *result = (PresenceDiscoveryResult*)malloc(sizeof(PresenceDiscoveryResult));
  PresenceDiscoveryResult* result = new PresenceDiscoveryResult();
  result->medium = medium;
  return result;
}

void presence_discovery_result_add_action(PresenceDiscoveryResult* result,
                                          int action) {
  result->device.actions.push_back(action);
}